File: Engine/Android/core/src/test/kotlin/despia/engine/ConformanceTest.kt
package despia.engine

import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File
import kotlin.test.assertTrue

/**
 * The cross-runtime conformance runner — executes every fixture in
 * OpenSource/Conformance/jse (every .json file) ({scope, expression, expected} triples) through this
 * runtime's JSE and asserts the result with JSE's own equality (the same coercion table every
 * platform ships). The corpus is platform-count-agnostic (/web/10 W0): Swift is the reference
 * and regenerates it in record mode on Codemagic; Kotlin runs it here on every local build;
 * the TS kernel joins as runtime #3. A JSE change is illegal without a green corpus on every
 * runtime that ships it (/web/07).
 *
 * Fixture location: found by walking up from the working directory to the repo root
 * (the directory containing OpenSource/Conformance). Missing corpus = loud failure —
 * a silently-skipped conformance suite is how drift starts.
 */
class ConformanceTest {

    private fun corpusDir(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/jse")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/jse not found walking up from ${System.getProperty("user.dir")}")
        }
    }

    @TestFactory
    fun jseCorpus(): List<DynamicTest> {
        val files = corpusDir().listFiles { f -> f.extension == "json" }?.sorted() ?: emptyList()
        require(files.isNotEmpty()) { "conformance corpus is empty — OpenSource/Conformance/jse/*.json" }
        return files.flatMap { file ->
            val doc = json(file.readText()).foundationValue as? Map<*, *>
                ?: error("${file.name}: not a JSON object")
            @Suppress("UNCHECKED_CAST")
            val cases = doc["cases"] as? List<Map<String, Any?>>
                ?: error("${file.name}: no cases[]")
            cases.map { c ->
                val name = "${file.nameWithoutExtension}/${c["name"]}"
                DynamicTest.dynamicTest(name) {
                    @Suppress("UNCHECKED_CAST")
                    val scope = (c["scope"] as? Map<String, Any?>) ?: emptyMap()
                    val store = StackStore()
                    scope.forEach { (k, v) -> store.vars[k] = v }
                    val actual = JSE.eval(c["expression"] as String, store, null)
                    val expected = c["expected"]
                    assertTrue(
                        JSE.equals(actual, expected),
                        "$name: expression `${c["expression"]}` -> ${JSE.string(actual)} (expected ${JSE.string(expected)})"
                    )
                }
            }
        }
    }
}

File: Web/packages/kernel/test/conformance.test.ts
//
//  conformance.test.ts - the cross-runtime conformance runner (TS twin of
//  Engine/Android ConformanceTest.kt). Executes every fixture in
//  OpenSource/Conformance/jse through this runtime's JSE and asserts with JSE's own
//  equality. Runs BOTH TS paths — the interpreter AND the compiled-JS closures — per
//  /web/07 ("the conformance corpus runs against BOTH TS paths").
//
//  Missing corpus = loud failure — a silently-skipped conformance suite is how drift
//  starts.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readdirSync, readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import { JSE, StackStore } from "../src/jse/jse.ts";
import { evalCompiled } from "../src/compile/codegen.ts";
import { string, NSNull, isDict, type Dict } from "../src/jse/values.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/jse");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error(`OpenSource/Conformance/jse not found walking up from ${import.meta.dirname}`);
    dir = parent;
  }
}

/** JSON → JSE values: explicit nulls become the present-null sentinel (the twins'
 *  json() mapping), numbers are doubles already. */
function toJseValue(v: unknown): unknown {
  if (v === null) return NSNull;
  if (Array.isArray(v)) return v.map(toJseValue);
  if (typeof v === "object") {
    const out: Dict = {};
    for (const [k, val] of Object.entries(v as Dict)) out[k] = toJseValue(val);
    return out;
  }
  return v;
}

type Case = { name: string; scope: Dict; expression: string; expected: unknown };

const files = readdirSync(corpusDir()).filter((f) => f.endsWith(".json")).sort();
assert.ok(files.length > 0, "conformance corpus is empty — OpenSource/Conformance/jse/*.json");

for (const file of files) {
  const doc = JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as { cases?: Case[] };
  const cases = doc.cases;
  assert.ok(Array.isArray(cases), `${file}: no cases[]`);
  for (const c of cases!) {
    const expected = toJseValue(c.expected ?? null);
    const seed = (): StackStore => {
      const store = new StackStore();
      for (const [k, v] of Object.entries(c.scope ?? {})) store.vars.set(k, toJseValue(v));
      return store;
    };
    test(`${file.replace(/\.json$/, "")}/${c.name} [interpreter]`, () => {
      const actual = JSE.eval(c.expression, seed(), null);
      assert.ok(
        JSE.equals(actual, expected),
        `expression \`${c.expression}\` -> ${string(actual)} (expected ${string(expected)})`,
      );
    });
    test(`${file.replace(/\.json$/, "")}/${c.name} [compiled]`, () => {
      const actual = evalCompiled(c.expression, seed(), null);
      assert.ok(
        JSE.equals(actual, expected),
        `expression \`${c.expression}\` -> ${string(actual)} (expected ${string(expected)})`,
      );
    });
  }
}

// keep the import "used" for erasable-syntax type stripping
void isDict;
